iwyu.cc

```diff
@@ -2594,8 +2583,17 @@ class IwyuBaseAstVisitor : public BaseAstVisitor<Derived> {
     vector<const UsingDecl*> using_decls
         = FindInMultiMap(visitor_state_->using_declarations, decl);
     for (Each<const UsingDecl*> it(&using_decls); !it.AtEnd(); ++it) {
-      if (!(*it)->getDeclContext()->Encloses(using_context))
-        continue;
+      const DeclContext* using_decl_context = (*it)->getDeclContext();
+      if (!using_decl_context->Encloses(use_context)) {
```

Comment:
The original code failed to find the using decl of a symbol if the using decl is in side a function.  For example;

```cpp
void foo() {
  using std::swap;
  std::vector a, b;
  swap(a, b);
}
```
In this example, GetUsingDeclarationOf 'swap' returns NULL, which is not what we want.  As far as I can tell, 
```cpp
(!(*it)->getDeclContext()->Encloses(using_context))
```
returns false whenever the using decl is in the same function it is used in and I don't know why.  This is what I meant by a questionable check.  But, as noted in the comment, in practise, it doesn't matter because we only care about the encloses case to make sure we're not finding a using decl across namespaces.

I think....